File: Flask Migration/utils/data_processor.py
"""
Data processing utilities for the Restaurant Ingredient Tracker.
"""

import pandas as pd
import numpy as np
from typing import Dict, List, Optional, Tuple


class DataProcessor:
    """Handles all data processing operations for ingredient tracking."""
    
    def __init__(self):
        """Initialize the data processor."""
        self.required_columns = {
            'ingredient_info': ['Ingredient', 'Unit Cost'],
            'input_stock': ['Ingredient', 'Received Qty'],
            'usage': ['Ingredient', 'Used Qty'],
            'waste': ['Ingredient', 'Wasted Qty']
        }
    
    def validate_csv_structure(self, file_path: str, file_type: str) -> bool:
        """Validate that CSV has required columns."""
        try:
            df = pd.read_csv(file_path)
            required_cols = self.required_columns[file_type]
            missing_cols = [col for col in required_cols if col not in df.columns]
            
            if missing_cols:
                raise ValueError(f"Missing columns in {file_type}: {missing_cols}")
            
            return True
        except Exception as e:
            raise ValueError(f"Error validating {file_type}: {str(e)}")
    
    def process_files(self, file_paths: Dict[str, str]) -> pd.DataFrame:
        """Process uploaded CSV files and calculate metrics."""
        
        # Validate all files first
        for file_type, file_path in file_paths.items():
            self.validate_csv_structure(file_path, file_type)
        
        # Load data
        ingredient_info = pd.read_csv(file_paths['ingredient_info'])
        input_stock = pd.read_csv(file_paths['input_stock'])
        usage = pd.read_csv(file_paths['usage'])
        waste = pd.read_csv(file_paths['waste'])
        
        # Clean and standardize data
        for df in [ingredient_info, input_stock, usage, waste]:
            df['Ingredient'] = df['Ingredient'].str.strip().str.title()
        
        # Merge all data on ingredient
        merged = ingredient_info.copy()
        merged = merged.merge(input_stock, on='Ingredient', how='left')
        merged = merged.merge(usage, on='Ingredient', how='left')
        merged = merged.merge(waste, on='Ingredient', how='left')
        
        # Fill missing values with 0
        numeric_columns = ['Received Qty', 'Used Qty', 'Wasted Qty']
        for col in numeric_columns:
            merged[col] = pd.to_numeric(merged[col], errors='coerce').fillna(0)
        
        # Calculate metrics
        merged['Expected Use'] = merged['Used Qty'] + merged['Wasted Qty']
        merged['Shrinkage'] = merged['Received Qty'] - merged['Expected Use']
        
        # Calculate costs
        merged['Used Cost'] = merged['Used Qty'] * merged['Unit Cost']
        merged['Waste Cost'] = merged['Wasted Qty'] * merged['Unit Cost']
        merged['Shrinkage Cost'] = merged['Shrinkage'] * merged['Unit Cost']
        merged['Total Cost'] = merged['Used Cost'] + merged['Waste Cost'] + merged['Shrinkage Cost']
        
        # Calculate percentages
        merged['Waste %'] = (merged['Wasted Qty'] / merged['Received Qty'] * 100).fillna(0)
        merged['Shrinkage %'] = (merged['Shrinkage'] / merged['Received Qty'] * 100).fillna(0)
        
        # Round numerical columns
        numeric_cols = ['Unit Cost', 'Used Cost', 'Waste Cost', 'Shrinkage Cost', 'Total Cost']
        for col in numeric_cols:
            merged[col] = merged[col].round(2)
        
        percentage_cols = ['Waste %', 'Shrinkage %']
        for col in percentage_cols:
            merged[col] = merged[col].round(1)
        
        return merged
    
    def apply_filters(self, data: pd.DataFrame, filter_type: str) -> pd.DataFrame:
        """Apply filters to the data."""
        if filter_type == 'high_shrinkage':
            return data[data['Shrinkage Cost'] > 10]
        elif filter_type == 'high_waste':
            return data[data['Waste %'] > 5]
        elif filter_type == 'missing_stock':
            return data[data['Received Qty'] == 0]
        elif filter_type == 'negative_shrinkage':
            return data[data['Shrinkage'] < 0]
        else:
            return data
    
    def sort_results(self, data: pd.DataFrame, sort_by: str, order: str = 'desc') -> pd.DataFrame:
        """Sort results by specified column."""
        ascending = order == 'asc'
        return data.sort_values(by=sort_by, ascending=ascending)
    
    def calculate_summary_stats(self, data: pd.DataFrame) -> Dict:
        """Calculate summary statistics."""
        return {
            'total_ingredients': len(data),
            'total_cost': data['Total Cost'].sum(),
            'total_waste_cost': data['Waste Cost'].sum(),
            'total_shrinkage_cost': data['Shrinkage Cost'].sum(),
            'avg_waste_percentage': data['Waste %'].mean(),
            'avg_shrinkage_percentage': data['Shrinkage %'].mean(),
            'high_shrinkage_items': len(data[data['Shrinkage Cost'] > 10]),
            'missing_stock_items': len(data[data['Received Qty'] == 0])
        }
    
    def get_alerts(self, data: pd.DataFrame) -> List[Dict]:
        """Generate alerts for problematic items."""
        alerts = []
        
        # High shrinkage alerts
        high_shrinkage = data[data['Shrinkage Cost'] > 50]
        for _, item in high_shrinkage.iterrows():
            alerts.append({
                'type': 'high_shrinkage',
                'severity': 'critical',
                'message': f"Critical shrinkage: {item['Ingredient']} has ${item['Shrinkage Cost']:.2f} in missing inventory",
                'ingredient': item['Ingredient'],
                'value': item['Shrinkage Cost']
            })
        
        # High waste alerts
        high_waste = data[data['Waste %'] > 15]
        for _, item in high_waste.iterrows():
            alerts.append({
                'type': 'high_waste',
                'severity': 'warning',
                'message': f"High waste: {item['Ingredient']} has {item['Waste %']:.1f}% waste rate",
                'ingredient': item['Ingredient'],
                'value': item['Waste %']
            })
        
        # Missing stock alerts
        missing_stock = data[data['Received Qty'] == 0]
        for _, item in missing_stock.iterrows():
            alerts.append({
                'type': 'missing_stock',
                'severity': 'warning',
                'message': f"No stock received for {item['Ingredient']} but usage/waste recorded",
                'ingredient': item['Ingredient'],
                'value': 0
            })
        
        return alerts
    
    def get_insights(self, data: pd.DataFrame) -> List[str]:
        """Generate insights from the data."""
        insights = []
        
        # Top cost contributors
        top_cost_items = data.nlargest(3, 'Total Cost')
        total_cost = data['Total Cost'].sum()
        
        for _, item in top_cost_items.iterrows():
            percentage = (item['Total Cost'] / total_cost) * 100
            insights.append(f"{item['Ingredient']} accounts for {percentage:.1f}% of total costs (${item['Total Cost']:.2f})")
        
        # Waste insights
        avg_waste = data['Waste %'].mean()
        if avg_waste > 10:
            insights.append(f"Average waste rate of {avg_waste:.1f}% is above recommended 10% threshold")
        
        # Shrinkage insights
        total_shrinkage = data['Shrinkage Cost'].sum()
        if total_shrinkage > 100:
            insights.append(f"Total shrinkage cost of ${total_shrinkage:.2f} indicates potential theft or inventory management issues")
        
        # Efficiency insights
        efficient_items = data[(data['Waste %'] < 5) & (data['Shrinkage %'].abs() < 5)]
        if len(efficient_items) > 0:
            insights.append(f"{len(efficient_items)} ingredients show excellent inventory management with low waste and shrinkage")
        
        return insights

File: Cloudflare Worker/src/utils/dataProcessor.js
/**
 * Data processing utilities for CSV analysis
 */

export async function processCSVData(csvData) {
  try {
    // Parse CSV data into objects
    const ingredientInfo = parseCSV(csvData.ingredient_info)
    const inputStock = parseCSV(csvData.input_stock)
    const usage = parseCSV(csvData.usage)
    const waste = parseCSV(csvData.waste)
    
    // Validate data structure
    validateCSVStructure(ingredientInfo, 'ingredient_info', ['Ingredient', 'Unit Cost'])
    validateCSVStructure(inputStock, 'input_stock', ['Ingredient', 'Received Qty'])
    validateCSVStructure(usage, 'usage', ['Ingredient', 'Used Qty'])
    validateCSVStructure(waste, 'waste', ['Ingredient', 'Wasted Qty'])
    
    // Create lookup maps for faster processing
    const stockMap = createLookupMap(inputStock, 'Ingredient', 'Received Qty')
    const usageMap = createLookupMap(usage, 'Ingredient', 'Used Qty')
    const wasteMap = createLookupMap(waste, 'Ingredient', 'Wasted Qty')
    
    // Process each ingredient
    const results = ingredientInfo.map(ingredient => {
      const name = ingredient.Ingredient
      const unitCost = parseFloat(ingredient['Unit Cost']) || 0
      const receivedQty = parseFloat(stockMap[name]) || 0
      const usedQty = parseFloat(usageMap[name]) || 0
      const wastedQty = parseFloat(wasteMap[name]) || 0
      
      // Calculate metrics
      const expectedUse = usedQty + wastedQty
      const shrinkage = receivedQty - expectedUse
      
      // Calculate costs
      const usedCost = usedQty * unitCost
      const wasteCost = wastedQty * unitCost
      const shrinkageCost = shrinkage * unitCost
      const totalCost = receivedQty * unitCost
      
      // Calculate percentages
      const wastePercentage = receivedQty > 0 ? (wastedQty / receivedQty) * 100 : 0
      const shrinkagePercentage = receivedQty > 0 ? (shrinkage / receivedQty) * 100 : 0
      
      return {
        Ingredient: name,
        'Unit Cost': unitCost,
        'Received Qty': receivedQty,
        'Used Qty': usedQty,
        'Wasted Qty': wastedQty,
        'Expected Use': expectedUse,
        Shrinkage: shrinkage,
        'Used Cost': usedCost,
        'Waste Cost': wasteCost,
        'Shrinkage Cost': shrinkageCost,
        'Total Cost': totalCost,
        'Waste %': wastePercentage,
        'Shrinkage %': shrinkagePercentage
      }
    })
    
    return results
    
  } catch (error) {
    console.error('Data processing error:', error)
    throw new Error(`Failed to process CSV data: ${error.message}`)
  }
}

function parseCSV(csvText) {
  const lines = csvText.trim().split('\n')
  if (lines.length < 2) {
    throw new Error('CSV must have at least a header and one data row')
  }
  
  const headers = lines[0].split(',').map(h => h.trim())
  const data = []
  
  for (let i = 1; i < lines.length; i++) {
    const values = lines[i].split(',').map(v => v.trim())
    if (values.length !== headers.length) {
      console.warn(`Row ${i + 1} has ${values.length} values but expected ${headers.length}`)
      continue
    }
    
    const row = {}
    headers.forEach((header, index) => {
      row[header] = values[index]
    })
    data.push(row)
  }
  
  return data
}

function validateCSVStructure(data, fileName, requiredColumns) {
  if (!data || data.length === 0) {
    throw new Error(`${fileName} is empty or invalid`)
  }
  
  const actualColumns = Object.keys(data[0])
  const missingColumns = requiredColumns.filter(col => !actualColumns.includes(col))
  
  if (missingColumns.length > 0) {
    throw new Error(`${fileName} is missing required columns: ${missingColumns.join(', ')}`)
  }
}

function createLookupMap(data, keyColumn, valueColumn) {
  const map = {}
  data.forEach(row => {
    map[row[keyColumn]] = row[valueColumn]
  })
  return map
}

export function calculateSummaryStats(results) {
  if (!results || results.length === 0) {
    return {
      totalIngredients: 0,
      totalCost: 0,
      totalWasteCost: 0,
      totalShrinkageCost: 0,
      averageWastePercentage: 0,
      averageShrinkagePercentage: 0
    }
  }
  
  const totalIngredients = results.length
  const totalCost = results.reduce((sum, item) => sum + (item['Total Cost'] || 0), 0)
  const totalWasteCost = results.reduce((sum, item) => sum + (item['Waste Cost'] || 0), 0)
  const totalShrinkageCost = results.reduce((sum, item) => sum + (item['Shrinkage Cost'] || 0), 0)
  
  const averageWastePercentage = results.reduce((sum, item) => sum + (item['Waste %'] || 0), 0) / totalIngredients
  const averageShrinkagePercentage = results.reduce((sum, item) => sum + (item['Shrinkage %'] || 0), 0) / totalIngredients
  
  return {
    totalIngredients,
    totalCost: Math.round(totalCost * 100) / 100,
    totalWasteCost: Math.round(totalWasteCost * 100) / 100,
    totalShrinkageCost: Math.round(totalShrinkageCost * 100) / 100,
    averageWastePercentage: Math.round(averageWastePercentage * 100) / 100,
    averageShrinkagePercentage: Math.round(averageShrinkagePercentage * 100) / 100
  }
}

export function getInsights(results) {
  if (!results || results.length === 0) {
    return []
  }
  
  const insights = []
  
  // High waste items
  const highWasteItems = results.filter(item => (item['Waste %'] || 0) > 15)
  if (highWasteItems.length > 0) {
    insights.push(`High waste detected: ${highWasteItems.length} items have >15% waste`)
  }
  
  // High shrinkage items
  const highShrinkageItems = results.filter(item => (item['Shrinkage Cost'] || 0) > 10)
  if (highShrinkageItems.length > 0) {
    insights.push(`Significant shrinkage: ${highShrinkageItems.length} items have >$10 shrinkage cost`)
  }
  
  // Missing stock items
  const missingStockItems = results.filter(item => (item['Received Qty'] || 0) === 0)
  if (missingStockItems.length > 0) {
    insights.push(`Missing stock: ${missingStockItems.length} items not received`)
  }
  
  return insights
}

export function filterResults(results, filterType) {
  if (!results || filterType === 'all') {
    return results
  }
  
  switch (filterType) {
    case 'high_waste':
      return results.filter(item => (item['Waste %'] || 0) > 10)
    case 'high_shrinkage':
      return results.filter(item => (item['Shrinkage Cost'] || 0) > 10)
    case 'missing_stock':
      return results.filter(item => (item['Received Qty'] || 0) === 0)
    case 'profitable':
      return results.filter(item => (item['Shrinkage Cost'] || 0) <= 0 && (item['Waste %'] || 0) <= 5)
    default:
      return results
  }
}

export function sortResults(results, sortBy, sortOrder = 'desc') {
  if (!results || results.length === 0) {
    return results
  }
  
  const columnMapping = {
    'shrinkage_cost': 'Shrinkage Cost',
    'total_cost': 'Total Cost',
    'waste_cost': 'Waste Cost',
    'used_cost': 'Used Cost',
    'waste_percentage': 'Waste %',
    'shrinkage_percentage': 'Shrinkage %',
    'received_qty': 'Received Qty',
    'used_qty': 'Used Qty',
    'wasted_qty': 'Wasted Qty',
    'ingredient': 'Ingredient',
    'unit_cost': 'Unit Cost'
  }
  
  const columnName = columnMapping[sortBy] || sortBy
  
  return results.sort((a, b) => {
    let aVal = a[columnName]
    let bVal = b[columnName]
    
    // Handle string vs number comparison
    if (typeof aVal === 'string' && typeof bVal === 'string') {
      aVal = aVal.toLowerCase()
      bVal = bVal.toLowerCase()
    } else {
      aVal = parseFloat(aVal) || 0
      bVal = parseFloat(bVal) || 0
    }
    
    if (sortOrder === 'asc') {
      return aVal > bVal ? 1 : (aVal < bVal ? -1 : 0)
    } else {
      return aVal < bVal ? 1 : (aVal > bVal ? -1 : 0)
    }
  })
}